### DISCUSSION TITLE
Tab 组件性能讨论

### DISCUSSION THREAD

**Discussion by Shiki-ARIA** [youzan/vant#12335] [General]
### What problem does this feature solve?

希望官方大大们能优化下tab组件的性能

### What does the proposed API look like?

1、双方组件都启用了手势滑动切换，
2、vant的tab组件是静态内容，且渲染的dom节点很少，Nut-ui的tab组件渲染的是200个列表的动态数据。但是通过性能对比可以看到，vant的tab组件消耗的性能远大于Nut-ui。并且经过我实测，vant的tab组件去渲染这动态的200个列表数据，切换的时候整个动画效果很卡很卡，而Nut-ui则依然很流畅。

![{9D7A40AE-B2B7-4dfa-B442-CD93156DCA8A}](https://github.com/youzan/vant/assets/83390614/878c8402-aea7-4339-808c-f6e6ae3e0348)

**Comment by chenjiahan** [comment]
试了下渲染 1000 个 Cell 组件还是挺流畅的，你可以提供一下对应的 codesandbox 链接吗？

**Reply by 931149046** [reply]
我这个tabs也很卡，https://mrylin.gitee.io/lovestatic/test/#/layouts

**Reply by 931149046** [reply]
使用真机测试，滑动切换很卡

**Reply by Shiki-ARIA** [reply]
https://ceshi.yblyr.cn/reception/test
这是我的测试tabs，真的真机上很卡

**Reply by Shiki-ARIA** [reply]
这是代码：
造成卡顿的原因也跟VanSpace这个组件有很大关系，去掉这个组件，性能会好很多。但是同样的代码用nutui的tab组件的话，不会卡（nutui的tab组件：https://ceshi.yblyr.cn/reception/nutest）。
![image](https://github.com/youzan/vant/assets/83390614/2faa80ce-612d-4153-8b41-034e54bd26ea)

**Reply by Shiki-ARIA** [reply]
vant版本是4.6.8

**Reply by 931149046** [reply]
![image](https://github.com/youzan/vant/assets/36146645/098942db-5625-48ad-801a-ea486da3dd53)

我没有vanspace也很卡

**Comment by chenjiahan** [comment]
尝试渲染 10 个 Tab 和 3000 个 VanCell 组件，并从第 1 个 Tab 滑动到第 10 个 Tab，渲染开销如下：

### Vant

<img width="449" alt="3d" src="https://github.com/youzan/vant/assets/7237365/bf0bfd01-4a5e-4583-ae23-d402a94d547f">

### Nut UI

<img width="442" alt="Screenshot 2023-10-03 at 21 39 21" src="https://github.com/youzan/vant/assets/7237365/fe2139dc-c8de-4e2b-b7d2-d22c27a3040c">

### 结论

在本 case 下，Vant Tabs 的 rendering 性能相对好一些，未能复现你描述的性能问题。

**Reply by Shiki-ARIA** [reply]
https://ceshi.yblyr.cn/reception/test
这是我的测试tabs，真的真机上很卡

**Reply by zhousg** [reply]
我试了一下没有感觉到卡

**Reply by 931149046** [reply]
我的很卡啊

**Reply by Shiki-ARIA** [reply]
请问你是在手机上测试的吗？如果是的话，是不是手机性能太好了？

**Reply by 931149046** [reply]
是真机上测的，https://mrylin.gitee.io/lovestatic/test/#/layouts 这个地址

**Reply by zhousg** [reply]
可以提供一个录屏看看~

**Reply by 931149046** [reply]
这上面我发不来视频啊，你可以测一下就知道了

**Reply by 931149046** [reply]
我感觉是浏览器兼容问题，我苹果手机微信和qq浏览器打开滑动切换测试就很卡，用华为自带浏览器滑动切换就不卡

**Comment by Shiki-ARIA** [comment]
我个人感觉似乎tabs组件在切换tab进行渲染动画的过程中，似乎还在做什么计算，而且似乎跟vue的双向绑定相关，导致触发了大量了patch事件，从而引起了css动画堵塞。我数了下性能分析中，每一帧的事件中，触发了至少约16次左右的vue3的patch事件，甚至有一帧触发了2次touchmove事件，从而触发了2*16次的 patch事件。这也是性能分析中每次切换tab的时候，cpu消耗性能大的原因吧？

**Reply by Shiki-ARIA** [reply]
1、下面的vant的Tab组件触发事件，由于触发的事件太多了，我这个屏幕显示不全，截图只是这一帧触发的事件一部分：
![image](https://github.com/youzan/vant/assets/83390614/d87b2151-c408-4442-8179-f0a59f9fb997)
2、下面的Nutui的Tab组件触发事件，Nutui每帧执行的事件就比vant的Tab组件触发的事件要少很多：
![image](https://github.com/youzan/vant/assets/83390614/c963acff-afb3-4e7f-99c5-aaa8143485e4)

**Comment by Shiki-ARIA** [comment]
这是我录制的视频，可以看到vant的tab组件在切换的时候，动画有明显的卡顿。
https://github.com/youzan/vant/assets/83390614/8aa6d1f1-2d58-4c70-b9c2-70347604b67a
